src/index.ts

```diff
@@ -1,21 +1,93 @@
+
+import {
+    INotebookTracker
+} from '@jupyterlab/notebook';
+
+import {
+  ContentsManager
+} from '@jupyterlab/services';
+
+import {
+  ICommandPalette
+} from '@jupyterlab/apputils';
+
 import {
   JupyterLab,
   JupyterLabPlugin
 } from '@jupyterlab/application';
 
+import {AllOfUsConfig} from './config';
+import {ConceptsWidget} from './concepts-widget';
+
 import '../style/index.css';
 
+// Activate the jupyterhub extension.
+function activateExtension(app: JupyterLab,
+                           notebooks: INotebookTracker,
+                           palette: ICommandPalette): void {
+    const tracker = notebooks;
+    let allOfUsConfig = new AllOfUsConfig();
+
+    // Create a single widget
+    let conceptsWidget: ConceptsWidget = new ConceptsWidget(allOfUsConfig);
+
+    // Add an application command
+    const conceptsCommand: string = 'allOfUs:concepts';
+    app.commands.addCommand(conceptsCommand, {
+      label: 'All of Us Concepts',
+      execute: () => {
+        if (!conceptsWidget.isAttached) {
+          // Attach the widget to the main area if it's not there
+          app.shell.addToMainArea(conceptsWidget, { mode: 'split-bottom' });
+        }
+        // Activate the widget
+        app.shell.activateById(conceptsWidget.id);
+      }
+    });
+
+
+    // Add the command to the palette.
+    palette.addItem({command: conceptsCommand, category: 'All Of Us'});
+
+    let initial_load = true;
+    tracker.currentChanged.connect(() =>
+    {
+      if (tracker.currentWidget) {
+        let path = tracker.currentWidget.context.path;
+        let slashIndex = path.lastIndexOf('/');
+        let directory = ''
+        if (slashIndex != -1) {
+          directory = path.substring(0, slashIndex + 1);
+        }
+        let contents = new ContentsManager();
+        contents.get(directory + '.all_of_us_config.json').then(
+              (model) => {
+                allOfUsConfig.setJson(model.content);
```

Comment:
Seems like we might want a model here where we're caching a few different instances of the widget (one per workspace). Seems like you wouldn't want the widget to be reset if you're switching between two notebooks in the same workspace. Separately, if I switch from workspace A -> B -> A, the widget should be reset on B (content may not be applicable), but would be nice if state persisted when returning to A.

Reply:
I agree that when nothing in the config has changed, nothing in the widget should reset. (Though there's nothing forcing the widget to do that just because they get notified.)

I added some logic here to check for this case so that we don't bother notifying widgets if nothing in the config has changed. It seems like if the config *does* change, it might be reasonable to reset the widget, or at least let the widgets decide if they want to reset based on whatever changed.

Reply:
Regarding having the state restored if you go from A -> B -> A...  that's not the UX I had in mind. I think it's easier to grok that there's a single extension on the page, and it's somewhat responsive to what notebook you have selected (some things in the widget will update when you switch notebooks)... but switching notebooks or workspaces shouldn't necessarily cause a total reset of it, and you don't necessarily expect its state to track what notebook you picked. 

Anyway, I think it's simpler for now to have a single widget and try it out with researchers before trying to keep around different widgets per notebook / workspace.

Reply:
Yeah I don't know if A->B->A is necessary or expected. I was imagining a case where I'm working in notebook A and playing with the widget, I flip focus to notebook B to reference how I did something in that notebook, then I come back. Suddenly my state is gone in the widget.

The reset I'd guess is necessary in most cases based on my understanding of the widget, but I probably need to see more concrete details before I can really know what behavior I'd expect.